onnxruntime/core/session/model_editor_c_api.cc

```diff
@@ -0,0 +1,358 @@
+    return OrtApis::CreateStatus(ORT_INVALID_ARGUMENT, "type_info cannot be null");
+  }
+
+  if (type_info->type != ONNX_TYPE_TENSOR) {
+    return OrtApis::CreateStatus(ORT_FAIL, "Only tensor types are supported currently");
+  }
+
+  if (type_info->tensor_type_info == nullptr) {
+    return OrtApis::CreateStatus(ORT_INVALID_ARGUMENT, "tensor_type_info cannot be null");
+  }
+
+  auto vi = std::make_unique<OrtValueInfo>();
+  vi->name = name;
+  vi->type_info = type_info->Clone();
+
+  *value_info = vi.release();
+
+  return nullptr;
+  API_IMPL_END
+}
+
+ORT_API_STATUS_IMPL(OrtModelEditorAPI::CreateNode, const char* operator_name, const char* domain_name,
+                    _In_ const char* node_name,
+                    _In_reads_(input_names_len) const char* const* input_names, size_t input_names_len,
+                    _In_reads_(output_names_len) const char* const* output_names, size_t output_names_len,
+                    _In_reads_(attribs_len) _Inout_opt_ OrtOpAttr** attributes, _In_opt_ size_t attribs_len,
+                    _Outptr_ OrtNode** node) {
+  API_IMPL_BEGIN
+  auto n = std::make_unique<OrtNode>();
+  n->operator_name = operator_name;
+  n->domain_name = domain_name == kOnnxDomainAlias ? kOnnxDomain : domain_name;
+  n->node_name = node_name;
+
+  n->input_names.reserve(input_names_len);
+  for (size_t i = 0; i < input_names_len; ++i) {
+    n->input_names.push_back(input_names[i]);
+  }
+
+  n->output_names.reserve(output_names_len);
+  for (size_t i = 0; i < output_names_len; ++i) {
+    n->output_names.push_back(output_names[i]);
+  }
+
+  if (attributes != nullptr) {
+    n->attributes.reserve(attribs_len);
+    for (size_t i = 0; i < attribs_len; ++i) {
+      n->attributes.push_back(*reinterpret_cast<const ONNX_NAMESPACE::AttributeProto*>(attributes[i]));
+      // take ownership. as we took a copy that means releasing the original value
+      OrtApis::ReleaseOpAttr(attributes[i]);
+      attributes[i] = nullptr;
+    }
+  }
+
+  *node = n.release();
+  return nullptr;
+  API_IMPL_END
+}
+
+ORT_API_STATUS_IMPL(OrtModelEditorAPI::CreateGraph, _Outptr_ OrtGraph** graph) {
+  API_IMPL_BEGIN
+  auto g = std::make_unique<OrtGraph>();
+
+  // do some reserves to reduce reallocation. if we had a hint about sizes upfront that would be optimal
+  g->inputs.reserve(8);
+  g->outputs.reserve(8);
+  g->initializers.reserve(32);
+  g->external_initializers.reserve(32);
+  g->nodes.reserve(64);
+
+  *graph = g.release();
+  return nullptr;
+  API_IMPL_END
+}
+
+ORT_API_STATUS_IMPL(OrtModelEditorAPI::SetGraphInputs, _In_ OrtGraph* graph,
+                    _In_reads_(inputs_len) _In_ OrtValueInfo** inputs, _In_ size_t inputs_len) {
+  API_IMPL_BEGIN
+  for (size_t i = 0; i < inputs_len; ++i) {
+    if (inputs[i] == nullptr) {
+      return OrtApis::CreateStatus(ORT_INVALID_ARGUMENT, "inputs cannot contain null entries");
+    }
+
+    graph->inputs.push_back(std::unique_ptr<OrtValueInfo>(inputs[i]));  // take ownership
```

Comment:
>std::unique_ptr<OrtValueInfo>(inputs[i])) [](http://example.com/codeflow?start=28&length=41)

This would be OK if the objects were not owned by anything prior to this, which is not OK from the clients point of view.

If client code had the ownership, it would have to relinquish it. Is any way to make this automatic?

Would making a copy be better here?

Does static analysis allow not using make_unique()?